### ISSUE TITLE
Kubernetes on Canonical distribution with Juju non working properly on vSphere

### ISSUE THREAD

**Issue by cgar-phot** [juju-solutions/bundle-canonical-kubernetes#480]
When deploying Kubernetes on Canonical distribution with Juju on vSphere there is a name resolution issue when adding a new kubernetes worker.
The sequence of steps that lead to the inconsistency is the following:

1) A new worker is requested ("juju add-unit kubernetes-worker")
2) The VM is created and start booting, the networking service aren't still started yet
3) The hostname is initially set as "ubuntu"
4) Networking is started up and a DHCP request for an IP lease is made, using hostname "ubuntu" (thus the DHCP server record the lease as "ubuntu")
5) The hostname is changed to proper value (ie: "juju-fe12e1-11"), but no additional DHCP request is sent to update the lease information, leaving it indefinitely as "ubuntu"

So basically on the DHCP server all the IP leases have the same "ubuntu" assigned name, and the same happens on the local DNS server if automatic update of records by DHCP server is enabled.

Manually restarting the VMs once fixes the situation, because at second boot the hostname is already correct and the DHCP registration has the correct "juju-xxxxxx-xx" name (the same would happens also if a dhclient release/renew command would be issued without actual rebooting). The trouble is that until either one of these actions is taken, the node is unresolvable by DNS name and that also cause Kubernetes issues.

**Comment by marcoceppi** [comment]
This is honestly an issue with Juju. As it's responsible for hostname and
day 0 configuration.

**Comment by CalvinHartwell** [comment]
+1 to @marcoceppi this is not a CDK issue, but with juju, it would be better to open a bug via the juju issue tracker [https://github.com/juju/juju/blob/develop/CONTRIBUTING.md](https://github.com/juju/juju/blob/develop/CONTRIBUTING.md). If you need help doing this let me know. 

@cgar-phot your DHCP server is really configured to give DHCP leases to machines based on hostname? Usually it is done based on MAC address, which should be pretty much guaranteed unique. 

This is pretty interesting, I am pretty sure we will hit this same issue today as well, so we will work on it.

**Comment by cgar-phot** [comment]
@CalvinHartwell 
no, our DHCP server doesn't give leases based on hostname, neither on MAC address.
It just create the lease with the name declared by DHCP client ("ubuntu" in that case) and create the corresponding "A record" in the DNS server (again with "ubuntu" as hostname).

**Comment by CalvinHartwell** [comment]
@cgar-phot we are also hitting a similar issue, it appears to be an issue with cloud-init according to engineering.

**Comment by galgalesh** [comment]
We're also hitting against this issue. This is a regression for us. cloud-init used to have the correct behavior, but it changed somewhere in the past few months.

**Comment by cgar-phot** [comment]
@CalvinHartwell do the existence of a workaround has been confirmed by your engineering ?
if positive, could you please post it here ?

**Comment by galgalesh** [comment]
We found a workaround: if you set the default lease time for the dhcp server to 5 minutes, the clients will renew the dhcp after 2.5 minutes with the correct hostname. This means that the DNS issue will be fixed before Kubernetes is completely installed.
